usr.bin/db/db.1
.\" 	$NetBSD$
.\"
.\" Copyright (c) 2002-2009,2012 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Luke Mewburn of Wasabi Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd February 17, 2012
.Dt DB 1
.Os
.
.Sh NAME
.Nm db
.Nd
manipulate
.Xr db 3 Ns 's
.Xr btree 3
and
.Xr hash 3
databases
.
.Sh SYNOPSIS
.Nm
.Op Fl DKiNqV
.Op Fl E Ar endian
.Op Fl f Ar infile
.Op Fl O Ar outsep
.Op Fl S Ar visitem
.Op Fl T Ar visspec
.Op Fl U Ar unvisitem
.Op Fl X Ar extravis
.Ar type
.Ar dbfile
.Op Ar key Op Ar \&.\&.\&.
.
.Nm
.Fl d
.Op Fl iNq
.Op Fl E Ar endian
.Op Fl f Ar infile
.Op Fl T Ar visspec
.Op Fl U Ar unvisitem
.Ar type
.Ar dbfile
.Op Ar key Op Ar \&.\&.\&.
.
.Nm
.Fl w
.Op Fl CDiNqR
.Op Fl E Ar endian
.Op Fl F Ar insep
.Op Fl f Ar infile
.Op Fl m Ar mode
.Op Fl P Ar pagesize
.Op Fl T Ar visspec
.Op Fl U Ar unvisitem
.Ar type
.Ar dbfile
.Op Ar key Ar value Op Ar \&.\&.\&.
.
.Sh DESCRIPTION
.Nm
allows manipulation of
.Xr btree 3
and
.Xr hash 3
.Pq Xr db 3
databases.
.Pp
.Nm
has three modes of operation to perform upon
.Ar dbfile :
.Bl -tag -width "delete" -offset indent
.It read
Displays the given
.Ar key Ns s ,
and keys described in
.Ar infile .
If no keys and no
.Ar infile
is specified, the entire database is displayed.
This is the default mode of operation.
.It delete
Enabled with
.Fl d .
Deletes the given
.Ar key Ns s ,
and keys described in
.Ar infile .
.It write
Enabled with
.Fl w .
Writes the given
.Ar key Ns s
and
.Ar value Ns s ,
and keys and values described in
.Ar infile
(in the latter case, entries are separated by
.Ar insep ) .
.El
.Pp
There are two mandatory arguments:
.Ar type
is the database type; either
.Sq Sy btree
or
.Sq Sy hash ,
and
.Ar dbfile
is the database file to manipulate.
.Pp
Options valid for all modes are:
.Bl -tag -width Fl -offset indent
.
.It Fl E Ar endian
Set the endianness of the database.
.Ar endian
may be one of:
.Bl -tag -width 1n -offset indent -compact
.It Sy B
Big endian
.It Sy H
Host endian
.It Sy L
Little endian
.El
Defaults to
.Sq H
(host endian).
.
.It Fl f Ar infile
Contains a list of keys
(for read and delete), or
.Ar insep
separated keys and values (for write)
to be used as arguments to the given mode.
If
.Ar infile
is
.Sq Sy - ,
.Dv stdin
is used.
.
.It Fl i
Keys are converted to lower case before manipulation.
.
.It Fl N
Include the NUL byte at the end of the key or value.
.Pp
By convention, each key and value in the underlying database is
terminated with a NUL byte, although it is possible to have keys
or values that are not so terminated.
Without the
.Fl N
option,
.Nm
adds a NUL terminator to any keys and values specified on the
command line or in the input file,
before searching or modifying the underlying database,
and removes the last byte (which is usually a NUL terminator)
before printing any keys and values retrieved from the
underlying database.
With the
.Fl N
option,
.Nm
does not add or remove a terminating byte.
The
.Fl S , Fl T ,
and
.Fl U
options may be useful to encode the NUL terminator in a readable form.
.
.It Fl q
Quiet operation.
In read mode, missing keys are not considered to be an error.
In delete
.Pq Fl d
and write
.Pq Fl w
modes,
the result of various operations is suppressed.
.
.It Fl T Ar visspec
Control how the items specified by the
.Fl S
option are encoded
and
.Fl U
option are decoded.
The
.Ar visspec
option-argument is a string specifying
.Xr strsvisx 3
flags.
The string consists of one or more characters:
.Bl -tag -width 2n -offset indent -compact
.It Sy b
.Dv VIS_NOSLASH
.It Sy c
.Dv VIS_CSTYLE .
Overrides
.Sy h
and
.Sy o .
.It Sy h
.Dv VIS_HTTPSTYLE .
Overrides
.Sy c
and
.Sy o .
.It Sy o
.Dv VIS_OCTAL .
Overrides
.Sy c
and
.Sy h .
.It Sy s
.Dv VIS_SAFE
.It Sy t
.Dv VIS_TAB
.It Sy w
.Dv VIS_WHITE
.El
.Pp
.Fl U
only supports
.Fl T
.Sy h .
.Pp
See
.Xr vis 1 Ns 's
corresponding options for the meaning of these characters,
and
.Xr strsvisx 3
for more detail on the flags.
.
.It Fl U Ar unvisitem
Specify items to
.Xr strunvisx 3
decode.
The
.Ar unvisitem
option-argument is a character specifying if the key
.Pq Ar k ,
the value
.Pq Ar v
or both
.Pq Ar b
should be decoded.
.
.El
.
.Pp
Read mode specific options are:
.
.Bl -tag -width Fl -offset indent
.
.It Fl D
Display duplicate entries in
.Sy btree
databases.
.
.It Fl K
Display key.
.
.It Fl O Ar outsep
Field separator string between key and value.
Defaults to a single tab
.Pq Sq \et .
.
.It Fl S Ar visitem
Specify items to
.Xr strsvisx 3
encode.
The
.Ar visitem
option-argument is a character specifying if the key
.Pq Ar k ,
the value
.Pq Ar v
or both
.Pq Ar b
should be encoded.
.
.It Fl V
Display value.
.
.It Fl X Ar extravis
When encoding items with
.Fl S
option also encode characters in
.Ar extravis ,
per
.Xr strsvisx 3 .
.
.El
.Pp
.Bd -ragged -offset indent
(If neither of
.Fl K
or
.Fl V
is given, both options are enabled.)
.Ed
.
.Pp
Write mode specific options are:
.Bl -tag -width Fl -offset indent
.
.It Fl C
Create new database, and truncate existing databases.
.
.It Fl D
Allow duplicate entries in
.Sy btree
databases.
(Requires
.Fl R
to be useful.)
.
.It Fl F Ar insep
Input field separator string between key and value used when parsing
.Ar infile .
Defaults to a single space
.Pq Sq \  .
.
.It Fl m Ar mode
Octal mode of created database.
Defaults to
.Sq 0644 .
.
.It Fl P Ar pagesize
Set the page size of the table to
.Ar pagesize
bytes.
If set to
.Sq 0 ,
a database-specific default is determined,
based on the block-size of the underlying file-system.
Defaults to
.Sq 4096 .
.
.It Fl R
Overwrite existing entries.
If not specified, writing to an existing entry raises an error.
.
.El
.
.Sh SEE ALSO
.Xr vis 1 ,
.Xr btree 3 ,
.Xr db 3 ,
.Xr hash 3 ,
.Xr strsvisx 3 ,
.Xr strunvisx 3
.\" .Xr makemap 8
.
.Sh HISTORY
The
.Nm
command appeared in
.Nx 2.0 .
.
.Sh AUTHORS
.An Luke Mewburn
.Aq lukem@NetBSD.org .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

usr.sbin/sdpd/db.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 2009 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Iain Hibbert.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__RCSID("$NetBSD$");

#include <bluetooth.h>
#include <sdp.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <uuid.h>

#include "sdpd.h"

/*
 * Using a prebuilt service record means that providing ServerState
 * and a non-hardcoded ProviderName are difficult. Look into that later.
 */

/* ServiceDiscoveryServer service record */
static uint8_t sds_data[] = {
	0x09, 0x00, 0x00,	//  uint16	ServiceRecordHandle
	0x0a, 0x00, 0x00, 0x00,	//  uint32	0x00000000
	0x00,

	0x09, 0x00, 0x01,	//  uint16	ServiceClassIDList
	0x35, 0x03,		//  seq8(3)
	0x19, 0x10, 0x00,	//   uuid16	ServiceDiscoveryServer

	0x09, 0x00, 0x04,	//  uint16	ProtocolDescriptorList
	0x35, 0x0d,		//  seq8(13)
	0x35, 0x06,		//   seq8(6)
	0x19, 0x01, 0x00,	//    uuid16	L2CAP
	0x09, 0x00, 0x01,	//    uint16	L2CAP_PSM_SDP
	0x35, 0x03,		//   seq8(3)
	0x19, 0x00, 0x01,	//    uuid16	SDP

	0x09, 0x00, 0x05,	//  uint16	BrowseGroupList
	0x35, 0x03,		//  seq8(3)
	0x19, 0x10, 0x02,	//   uuid16	PublicBrowseGroup

	0x09, 0x00, 0x06,	//  uint16	LanguageBaseAttributeIDList
	0x35, 0x09,		//  seq8(9)
	0x09, 0x65, 0x6e,	//   uint16	0x656e	("en")
	0x09, 0x00, 0x6a,	//   uint16	106	(UTF-8)
	0x09, 0x01, 0x00,	//   uint16	PrimaryLanguageBaseID

	0x09, 0x01, 0x00,	//  uint16	PrimaryLanguageBaseID + ServiceNameOffset
	0x25, 0x1b, 0x42, 0x6c,	//  str8(27)	"Bluetooth service discovery"
	0x75, 0x65, 0x74, 0x6f,
	0x6f, 0x74, 0x68, 0x20,
	0x73, 0x65, 0x72, 0x76,
	0x69, 0x63, 0x65, 0x20,
	0x64, 0x69, 0x73, 0x63,
	0x6f, 0x76, 0x65, 0x72,
	0x79,

	0x09, 0x01, 0x02,	//  uint16	PrimaryLanguageBaseID + ProviderNameOffset
	0x25, 0x06, 0x4e, 0x65,	//  str8(6)	"NetBSD"
	0x74, 0x42, 0x53, 0x44,

	0x09, 0x02, 0x00,	//  uint16	VersionNumberList
	0x35, 0x03,		//  seq8(3)
	0x09, 0x01, 0x00,	//   uint16	v1.0
};

/* BrowseGroupDescriptor service record */
static uint8_t bgd_data[] = {
	0x09, 0x00, 0x00,	//  uint16	ServiceRecordHandle
	0x0a, 0x00, 0x00, 0x00,	//  uint32	0x00000001
	0x01,

	0x09, 0x00, 0x01,	//  uint16	ServiceClassIDList
	0x35, 0x03,		//  seq8(3)
	0x19, 0x10, 0x01,	//   uuid16	BrowseGroupDescriptor

	0x09, 0x00, 0x06,	//  uint16	LanguageBaseAttributeIDList
	0x35, 0x09,		//  seq8(9)
	0x09, 0x65, 0x6e,	//   uint16	0x656e	("en")
	0x09, 0x00, 0x6a,	//   uint16	106	(UTF-8)
	0x09, 0x01, 0x00,	//   uint16	PrimaryLanguageBaseID

	0x09, 0x01, 0x00,	//  uint16	PrimaryLanguageBaseID + ServiceNameOffset
	0x25, 0x12, 0x50, 0x75,	//  str8(18)	"Public Browse Root"
	0x62, 0x6c, 0x69, 0x63,
	0x20, 0x42, 0x72, 0x6f,
	0x77, 0x73, 0x65, 0x20,
	0x52, 0x6f, 0x6f, 0x74,

	0x09, 0x02, 0x00,	//  uint16	GroupID
	0x19, 0x10, 0x02,	//  uuid16	PublicBrowseRoot
};

/*
 * Initialise the record database with the ServiceDiscoveryServer
 * and BrowseGroupDescriptor records
 */
bool
db_init(server_t *srv)
{
	sdp_data_t d;

	LIST_INIT(&srv->rlist);
	srv->handle = 0;

	d.next = sds_data;
	d.end = sds_data + sizeof(sds_data);
	if (!db_create(srv, -1, BDADDR_ANY, srv->handle++, &d))
		return false;

	d.next = bgd_data;
	d.end = bgd_data + sizeof(bgd_data);
	if (!db_create(srv, -1, BDADDR_ANY, srv->handle++, &d))
		return false;

	return true;
}

/*
 * Iterate through records selected by fd.  rec should point to a NULL
 * value to start the iteration, and false will be returned when there
 * are no more records to return.
 */
bool
db_next(server_t *srv, int fd, record_t **rec)
{
	record_t *r;

	if (*rec == NULL)
		r = LIST_FIRST(&srv->rlist);
	else
		r = LIST_NEXT(*rec, next);

	while (r != NULL && !FD_ISSET(fd, &r->refset))
		r = LIST_NEXT(r, next);

	*rec = r;
	return (r == NULL) ? false : true;
}

/*
 * Match a ServiceRecord against a UUID. Note that because we already
 * know that the record data is valid, we don't need to recurse here
 * and can just skip over SEQ and ALT headers. Return true if equivalent
 * UUID is found.
 */
static bool
db_match_uuid(record_t *rec, uuid_t *uuid)
{
	uint8_t *p = rec->data.next;
	uuid_t u;

	while (p < rec->data.end) {
		switch(*p++) {
		case SDP_DATA_NIL:
			break;

		case SDP_DATA_BOOL:
		case SDP_DATA_INT8:
		case SDP_DATA_UINT8:
		case SDP_DATA_SEQ8:
		case SDP_DATA_ALT8:
			p += 1;
			break;

		case SDP_DATA_INT16:
		case SDP_DATA_UINT16:
		case SDP_DATA_SEQ16:
		case SDP_DATA_ALT16:
			p += 2;
			break;

		case SDP_DATA_INT32:
		case SDP_DATA_UINT32:
		case SDP_DATA_SEQ32:
		case SDP_DATA_ALT32:
			p += 4;
			break;

		case SDP_DATA_INT64:
		case SDP_DATA_UINT64:
			p += 8;
			break;

		case SDP_DATA_INT128:
		case SDP_DATA_UINT128:
			p += 16;
			break;

		case SDP_DATA_STR8:
		case SDP_DATA_URL8:
			p += 1 + *p;
			break;

		case SDP_DATA_STR16:
		case SDP_DATA_URL16:
			p += 2 + be16dec(p);
			break;

		case SDP_DATA_STR32:
		case SDP_DATA_URL32:
			p += 4 + be32dec(p);
			break;

		case SDP_DATA_UUID16:
			u = BLUETOOTH_BASE_UUID;
			u.time_low = be16dec(p);

			if (uuid_equal(&u, uuid, NULL))
				return true;

			p += 2;
			break;

		case SDP_DATA_UUID32:
			u = BLUETOOTH_BASE_UUID;
			u.time_low = be32dec(p);

			if (uuid_equal(&u, uuid, NULL))
				return true;

			p += 4;
			break;

		case SDP_DATA_UUID128:
			uuid_dec_be(p, &u);

			if (uuid_equal(&u, uuid, NULL))
				return true;

			p += 16;
			break;

		default:
			return false;
		}
	}

	return false;
}

/*
 * Select ServiceRecords matching ServiceSearchPattern
 *
 * A record is selected when it is visible to the client and
 * contains each and every UUID from the ServiceSearchPattern
 */
void
db_select_ssp(server_t *srv, int fd, sdp_data_t *ssp)
{
	record_t *r;
	sdp_data_t s;
	uuid_t u;

	LIST_FOREACH(r, &srv->rlist, next) {
		if (!r->valid)
			continue;

		if (!srv->fdidx[fd].control
		    && !bdaddr_any(&r->bdaddr)
		    && !bdaddr_same(&r->bdaddr, &srv->fdidx[fd].bdaddr))
			continue;

		s = *ssp;
		for (;;) {
			if (!sdp_get_uuid(&s, &u)) {
				/* matched all UUIDs */
				FD_SET(fd, &r->refset);
				r->refcnt++;
				break;
			}

			if (!db_match_uuid(r, &u)) {
				/* does not match UUID */
				break;
			}
		}
	}
}

/*
 * Select a ServiceRecord given the RecordHandle.
 */
void
db_select_handle(server_t *srv, int fd, uint32_t handle)
{
	record_t *r;

	LIST_FOREACH(r, &srv->rlist, next) {
		if (!r->valid)
			continue;

		if (!srv->fdidx[fd].control
		    && !bdaddr_any(&r->bdaddr)
		    && !bdaddr_same(&r->bdaddr, &srv->fdidx[fd].bdaddr))
			continue;

		if (handle == r->handle) {
			FD_SET(fd, &r->refset);
			r->refcnt++;
			break;
		}
	}
}

/*
 * Create a record and insert in server record list in ascending handle
 * order. Where a selectable record exists with the same handle number,
 * it will be expired.
 */
bool
db_create(server_t *srv, int fd, const bdaddr_t *bdaddr, uint32_t handle, sdp_data_t *data)
{
	record_t *n, *r, *rec;
	sdp_data_t d, v;
	uint16_t a;
	size_t len;

	d = *data;
	if (!sdp_get_attr(&d, &a, &v)
	    || a != SDP_ATTR_SERVICE_RECORD_HANDLE
	    || sdp_data_type(&v) != SDP_DATA_UINT32)
		return false;

	sdp_set_uint(&v, handle);

	len = data->end - data->next;
	rec = malloc(sizeof(record_t) + len);
	if (rec == NULL)
		return false;

	memset(rec, 0, sizeof(record_t));
	FD_ZERO(&rec->refset);
	rec->handle = handle;
	rec->valid = true;
	rec->fd = fd;
	bdaddr_copy(&rec->bdaddr, bdaddr);
	rec->data.next = rec->ext;
	rec->data.end = rec->ext + len;
	memcpy(rec->ext, data->next, len);

	/*
	 * Note, this does not handle the case where we expire
	 * the first record on the list, as that won't happen.
	 */
	n = LIST_FIRST(&srv->rlist);
	if (n != NULL) {
		do {
			r = n;
			n = LIST_NEXT(r, next);
		} while (n != NULL && n->handle < handle);

		if (n != NULL && n->valid && n->handle == handle) {
			if (n->refcnt-- == 0) {
				LIST_REMOVE(n, next);
				free(n);
			} else {
				n->valid = false;
				n->fd = -1;
			}
		}

		LIST_INSERT_AFTER(r, rec, next);
	} else {
		LIST_INSERT_HEAD(&srv->rlist, rec, next);
	}

	return true;
}

/*
 * Unselect any ServiceRecords selected by fd
 */
void
db_unselect(server_t *srv, int fd)
{
	record_t *n, *r;

	n = LIST_FIRST(&srv->rlist);
	while (n != NULL) {
		r = n;
		n = LIST_NEXT(r, next);

		if (FD_ISSET(fd, &r->refset)) {
			if (r->refcnt-- == 0) {
				LIST_REMOVE(r, next);
				free(r);
			} else {
				FD_CLR(fd, &r->refset);
			}
		}
	}
}

/*
 * Invalidate or release all records owned by fd
 */
void
db_release(server_t *srv, int fd)
{
	record_t *n, *r;

	n = LIST_FIRST(&srv->rlist);
	while (n != NULL) {
		r = n;
		n = LIST_NEXT(r, next);

		if (r->fd == fd) {
			if (r->refcnt-- == 0) {
				LIST_REMOVE(r, next);
				free(r);
			} else {
				r->valid = false;
				r->fd = -1;
			}
		}
	}
}
